mcp/shared.go

```diff
@@ -333,6 +358,36 @@ func setProgressToken(p Params, pt any) {
 	m[progressTokenKey] = pt
 }
 
+// A Request is a method request with parameters and additional information, such as the session.
+// The parameters are untyped; see [ClientRequest] and [ServerRequest] for the typed versions.
+type Request[S Session] struct {
+	Session S
+	Params  Params
+}
+
+// A ClientRequest is a request to a client.
+type ClientRequest[P Params] struct {
+	Session *ClientSession
+	Params  P
+}
+
+// A ServerRequest is a request to a server.
+type ServerRequest[P Params] struct {
+	Session *ServerSession
+	Params  P
+}
+
+func (*ClientRequest[P]) isRequest() {}
```

Comment:
I don't think these are used, given that you settled on a concrete Request type.